### PR TITLE
TTS in Moshi

### PR DESCRIPTION
**Pull request by adefossez**
## Checklist

- [ ] Read CONTRIBUTING.md, and accept the CLA by including the provided snippet. We will not accept PR without this.
- [ ] Run pre-commit hook.
- [ ] If you changed Rust code, run `cargo check`, `cargo clippy`, `cargo test`.

## PR Description

Moves the implementation of the TTS inside core moshi, along with a new script to synthesize at once a number of things from a JSONL.

### CODE REVIEW

File: moshi/moshi/__init__.py [hunk 1]
```diff
@@ -16,4 +16,4 @@
 from . import quantization
 from . import utils
 
-__version__ = "0.2.5"
+__version__ = "0.2.6a1"
```
Comment (LaurentMazare) [inline moshi/moshi/__init__.py:19]: You probably want to pull from main, we already released 0.2.6.

File: moshi/moshi/models/tts.py [hunk 2]
```diff
@@ -0,0 +1,624 @@
...
+from sentencepiece import SentencePieceProcessor
+import sphn
+import torch
+
+from ..conditioners import ConditionAttributes, dropout_all_conditions, TensorCondition
+from ..conditioners.text import LUTConditioner
+from . import loaders, MimiModel, LMModel, LMGen
+
+
+DEFAULT_DSM_TTS_REPO = "alex-testing/dsm-tts-rc"
```
Comment (LaurentMazare) [inline moshi/moshi/models/tts.py:23]: Switch to the proper repo names under the kyutai org?

File: moshi/moshi/models/tts.py [hunk 3]
```diff
@@ -0,0 +1,624 @@
...
+            This is used for multiplexing multiple input tokens into the text stream.
+
+    """
+    word = 0
+    pad = 3
+    main = 1
+    other = 2
+    zero = -1
+    ungenerated = -2
+    card: int = 8001
```
Comment (LaurentMazare) [inline moshi/moshi/models/tts.py:47]: Maybe the card should come from a config as we're likely to tweak it for a potential TTS supporting more languages?

File: moshi/moshi/models/tts.py [hunk 4]
```diff
@@ -0,0 +1,624 @@
...
+@dataclass
+class Entry:
+    """One word to generate.
+
+    Args:
+        tokens: list of tokens for this word.
+        word: word as string.
+        padding: if > 0, we will force that many padding after the word.
+            To have an effect this should be more than len(tokens), as we will
+            prevent new word until we have passed all the text tokens.
```
Comment (LaurentMazare) [inline moshi/moshi/models/tts.py:59]: The wording here is a bit confusing to me, on one hand it says *padding after the word* but on the other end `len(tokens)` seems to be included as part of padding. Maybe this could be clarified?

File: moshi/moshi/models/tts.py [hunk 5]
```diff
@@ -0,0 +1,624 @@
...
+    entries: deque[Entry]
+    remaining_padding: int
+    forced_padding: int
+    queued: deque[int] = field(default_factory=deque)
+    lookahead: deque[int] = field(default_factory=deque)
+    end_step: int | None = None
+    consumption_times: list[int] = field(default_factory=list)
+    transcript: list[tuple[str, int]] = field(default_factory=list)
+
+    def get_tokens_ahead(self, lookahead: int) -> list[int]:
```
Comment (LaurentMazare) [inline moshi/moshi/models/tts.py:95]: Having both a parameter named `lookahead` and a `self.lookahead` field is a bit confusing to me, maybe you could rename these e.g. `lookahead_tokens`.

File: moshi/moshi/models/tts.py [hunk 6]
```diff
@@ -0,0 +1,624 @@
...
+                    return entry.tokens
+        return []
+
+
+def _delayed(codes: torch.Tensor, delays: list[int], fill_value: int) -> torch.Tensor:
+    # Apply the acoustic delay on the provided audio tokens.
+    K, T = codes.shape
+    out = torch.full((K, T + max(delays)), fill_value, device=codes.device, dtype=torch.long)
+    for k in range(K):
+        delay = delays[k]
```
Comment (LaurentMazare) [inline moshi/moshi/models/tts.py:110]: `enumerate(delays)`? (I guess `delays` should have the same len as `codes`)

File: moshi/moshi/models/tts.py [hunk 7]
```diff
@@ -0,0 +1,624 @@
...
+
+def _make_null(all_attributes: tp.Sequence[ConditionAttributes]) -> list[ConditionAttributes]:
+    # When using CFG, returns the null conditions.
+    return dropout_all_conditions(all_attributes)
+
+
+@dataclass
+class StateMachine:
+    """State machine that manipulates the `State` based on the model prediction.
+    In particular, every time the model predicts a `word` (see `TokenIds`) special token,
```
Comment (LaurentMazare) [inline moshi/moshi/models/tts.py:123]: Cosmetic but maybe it would be clearer to call the `word` token something like `new_word` (as word token seems a bit like a "regular" token).

File: moshi/moshi/models/tts.py [hunk 8]
```diff
@@ -0,0 +1,624 @@
...
+    the state machine will pop the next word to synthesize and start feeding it.
+    The model is optionally equipped with a second input text stream providing a lookahead
+    into the future text.
+
+    Args:
+        token_ids: special token values.
+        second_stream_ahead: if > 0, the model needs a second stream for lookahead.
+        max_padding: maximum number of padding token that can be sampled in a row.
```
Comment (LaurentMazare) [inline moshi/moshi/models/tts.py:131]: nit: `padding token` -> `padding tokens`

File: moshi/moshi/models/tts.py [hunk 9]
```diff
@@ -0,0 +1,624 @@
...
+                    # This allows us to pass the current word tokens faster.
+                    output = state.queued.popleft()
+                else:
+                    output = self.token_ids.pad
+            elif state.lookahead:
+                # Otherwise if we have some lookahead tokens we feed them.
+                second = state.lookahead.popleft()
+            # Then we multiplex the two tokens. We add `+1` to `second` so that
+            # we can encode -1, which would translate to an all 0s embedding.
+            output = (second + 1) * self.token_ids.card + output
```
Comment (LaurentMazare) [inline moshi/moshi/models/tts.py:239]: More of a modeling question here but why do we want to have the "dense" cross-product of the two tokens, wouldn't that necessitate a very large embedding table?

File: moshi/moshi/models/tts.py [hunk 10]
```diff
@@ -0,0 +1,624 @@
...
+    # break is indicated as e.g. <break time="3s"/>
+    event_re = re.compile(r"(?:<break\s+time=\"([0-9]+(?:.[0-9]*)?)s\"\s*/?>)|(?:\s+)")
+
+    def _add_entry(word: str):
+        nonlocal first_content, opened_main, last_speaker
+        assert ' ' not in word
+        assert word
+        tokens = tokenizer.encode(word)  # type: ignore
+        if first_content:
+            speaker = idx % len(speaker_tokens)
```
Comment (LaurentMazare) [inline moshi/moshi/models/tts.py:277]: Where is this `idx` defined? Probably this should be added as an `_add_entry` argument and used below?

File: moshi/moshi/models/tts.py [hunk 11]
```diff
@@ -0,0 +1,624 @@
...
+            first_content = False
+        padding = 0
+        if padding_between > 0:
+            padding = max(0, padding_between + len(tokens) - 1)
+        entries.append(Entry(tokens=tokens, word=word, padding=padding))
+
+    for idx, line in enumerate(script):
+        first_content = True
+        line = line.replace('’', "'")
+        line = line.replace(' : ', " ")
```
Comment (LaurentMazare) [inline moshi/moshi/models/tts.py:290]: I would have thought that we should either replace all the `:` with the empty string to get rid of all columns or we want to replace the French way of putting spaces `foo : bar` with the English way `foo: bar`?

File: moshi/moshi/models/tts.py [hunk 12]
```diff
@@ -0,0 +1,624 @@
...
+            prefixes: this should be the list of the lengths up until when to mask for the CFG.
+            cfg_is_no_prefix: if true, the null logits are computed with a masked prefix.
+            cfg_is_no_text: if true, the null logits are computed without the text.
+            **kwargs: passed to `moshi.models.lm.LMGen`.
+        """
+
+        def _main_wrapper(*args, **kwargs):
+            transformer_out, text_logits = original(*args, **kwargs)
+            if self.padding_bonus:
+                text_logits[..., 3] += self.padding_bonus
```
Comment (LaurentMazare) [inline moshi/moshi/models/tts.py:454]: Replace the 3 with `token_ids.pad`?

File: moshi/moshi/run_tts.py [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+        cfg_coef_conditioning = tts_model.cfg_coef
+        tts_model.cfg_coef = 1.
+        cfg_is_no_text = False
+        cfg_is_no_prefix = False
+    else:
+        cfg_is_no_text = True
+        cfg_is_no_prefix = True
+    mimi = tts_model.mimi
+
+    first_batch = True
```
Comment (LaurentMazare) [inline moshi/moshi/run_tts.py:98]: `first_batch` doesn't seem to be used anymore?

File: moshi/moshi/run_tts.py [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+            if prefixes is not None:
+                start_step = prefixes[idx].shape[-1]
+                start = int(mimi.sample_rate * start_step / mimi.frame_rate)
+                wav = wav[:, start:]
+            filename = args.out_folder / f"{request.id}.wav"
+            debug_tensors = {
+                'frames': frames[idx].short(),
+            }
+            sphn.write_wav(filename, wav.clamp(-1, 1).cpu().numpy(), mimi.sample_rate)
+            save_file(debug_tensors, filename.with_suffix('.safetensors'))
```
Comment (LaurentMazare) [inline moshi/moshi/run_tts.py:159]: Maybe saving the safetensors and json files should be optional and driven by a cli argument?

File: moshi/moshi/run_tts.py [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+        # However, some items might have finished earlier, in which case the computation for those
+        # was wasted. Effective speed accounts for that, and gives the speed up accounting for
+        # the actual amount of usable audio generated.
+        print(f"[TOT] Batch of size {len(batch)} took {time_taken:.2f}s, "
+              f"total speed {total_speed:.2f}x, "
+              f"effective speed {effective_speed:.2f}x")
+        batch.clear()
+
+    batch: list[TTSRequest] = []
+    with args.jsonl.open('r') as f:
```
Comment (LaurentMazare) [inline moshi/moshi/run_tts.py:195]: It would probably be nice to include a sample `jsonl` in the repo so as to have an easy way to try things out and know the expected format.

File: moshi/moshi/run_tts.py [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+    parser.add_argument("--nq", type=int, default=32, help="Number of codebooks to generate.")
+    parser.add_argument("--temp", type=float, default=0.6, help="Temperature for text and audio.")
+    parser.add_argument("--cfg-coef", type=float, default=2., help="CFG coefficient.")
+
+    parser.add_argument("--max-padding", type=int, default=8, help="Max padding in a row, in steps.")
+    parser.add_argument("--initial-padding", type=int, default=2, help="Initial padding, in steps.")
+    parser.add_argument("--final-padding", type=int, default=4, help="Amount of padding after the last word, in steps.")
+    parser.add_argument("--padding-bonus", type=float, default=0.,
+                        help="Bonus for the padding logits, should be between -2 and 2, "
+                             "will change the speed of speech.")
```
Comment (LaurentMazare) [inline moshi/moshi/run_tts.py:64]: Mention the direction of the impact? E.g. "a larger padding bonus causes the speech to be slower".

File: moshi/moshi/models/tts.py [hunk 17]
```diff
@@ -1,6 +1,11 @@
 # Copyright (c) Kyutai, all rights reserved.
 """
 Implements the logic for the state machine around the Delayed Streams Modeling (DSM) based TTS model.
+Things are more compelx than for STT models, where we can simply force feed the audio tokens and
```
Comment (LaurentMazare) [inline moshi/moshi/models/tts.py:4]: compelx -> complex
